### DISCUSSION TITLE
wg-easy without Masquerade?

### DISCUSSION THREAD

**Discussion by mellismike** [wg-easy/wg-easy#808] [General]
I'm trying to setup wg-easy without nat masquerade and I'm stuck.  I want the peers that connect to the server to retain the IP address in the subnet assigned when they connect to wireguard so that packets routed around my lan are using 10.11.0.0/24 and services can identify the wireguard peer.  I'm using NPM to reverse proxy on my lan and when masquerade is used all clients show up using the host IP.  I want to use NPM to whitelist/blacklist certain client IPs to only access certain services.

I used this tutorial [how-to-use-wireguard-without-masquarade](https://mattei.io/vpn/2022/01/08/how-to-use-wireguard-without-masquarade.html) and can't get it to work.  The guy who posted mentioned he using linuxservers image for wireguard.

The tutorial says not to use any post_up/post_down routing and I know wg-easy defaults to use nat masquerade if left blank.  I tried it two ways to get around this.
1) I used WG_POST_UP without the masquerade and some basic accept rules.
      - WG_POST_UP=iptables -A INPUT -p udp -m udp --dport 51820 -j ACCEPT;
                   iptables -A FORWARD -i wg0 -j ACCEPT;
                   iptables -A FORWARD -o wg0 -j ACCEPT;
2) I edited the wg0.conf directly in the container, removing the post_up rules, and issued the wg-quick down/up commands.  both did not work (and I know this approach won't work long term, since it gets overwritten on container restart).

Next I added a static route on my router.  UI edgerouter-x sfp.
![image](https://github.com/wg-easy/wg-easy/assets/135131238/7a504f78-1962-4b86-995f-33462f50ed0a)

- the host IP is 192.168.1.119
- the wireguard subnet is 10.11.0.0/24
- the peer is assigned 10.11.0.10

Here is my compose.yml and wg0.conf

```
services:
  wg-easy:
    environment:
      - WG_HOST=my.network.com
      - PASSWORD=${PASSWORD}
      - WG_PORT=51820
      - WG_DEFAULT_ADDRESS=10.11.0.x
      - WG_DEFAULT_DNS={dns}
      - WG_POST_UP=iptables -A INPUT -p udp -m udp --dport 51820 -j ACCEPT;
                   iptables -A FORWARD -i wg0 -j ACCEPT;
                   iptables -A FORWARD -o wg0 -j ACCEPT;
    image: ghcr.io/wg-easy/wg-easy:${TAG}
    container_name: wg-easy
    volumes:
      - .:/etc/wireguard
    ports:
      - "51820:51820/udp"
      - "51821:51821/tcp"
    restart: unless-stopped
    cap_add:
      - NET_ADMIN
      - SYS_MODULE
    sysctls:
      - net.ipv4.ip_forward=1
      - net.ipv4.conf.all.src_valid_mark=1
```

Server wg0.conf
```
[Interface]
PrivateKey = {private_key}
Address = 10.11.0.1/24
ListenPort = 51820
PreUp =
PostUp = iptables -A INPUT -p udp -m udp --dport 51820 -j ACCEPT; iptables -A FORWARD -i wg0 -j ACCEPT; iptables -A FORWARD -o wg0 -j ACCEPT;
PreDown =
PostDown =

# Client: my-client (999999-9999-9999-9999-999999999999)
[Peer]
PublicKey = {public_key}
PresharedKey = {preshared_key}
AllowedIPs = 10.11.0.10/32
```

Anyone have any ideas?

**Comment by mellismike** [comment]
If anyone is interested, I switch network_mode over to host and it started to work.  I'm not sure if there is a downside, but I guess it's like running wg directly on the host instead of inside a container as far as the network is concerned.  I'm guessing to get this to work using the container network I needed to figure out how to correctly route the incoming traffic back to the container.. maybe a static route on the host.  Not sure, but didn't get that far.

In the compose.yml file I had to remove the port mappings and sysctls.
and then added "network_mode: host".

Now, when wg peers access services on the lan their IP shows up in the range of 10.11.0.0/24, instead of the 192.168.1.119 (the host server).

**Reply by sigma2017** [reply]
Hello. Did you noticed a performance gain with this setup?
I set up wireguard directly on the host and it is running faster with ~ 50% but also with iptables -t nat -I POSTROUTING -s 10.10.10.0/24 -o enp0s3 -j MASQUERADE; 
I am curious to try your solution.

**Reply by mellismike** [reply]
It stopped working and I haven't had the time to try and troubleshoot it again.  When I get back around to it, I'm going to just use the host to see if this works instead of wg-easy.  I'll miss the great UI, but I'm not sure if this project is a good fit.

**Reply by sigma2017** [reply]
The UI is awesome, you can notice it was created by professionals. The main issue comes from the docker network bridge, compared to the native wireguard install it runs slower by 40% I would say based on my own tests. Either run wg-easy with network host (if possible) or better just install wireguard directly on the host, there are plenty of methods including easy way by using the linux package manager. Ofc, it will be much harder to configure it, you won't have the amazing UI but for sure it will be **much faster**. There are few unpolished solutions to make the management easier, I tested all, none of them do the right thing....I agree with you, wg-easy is wow but is not fast.

**Reply by eandersons** [reply]
@sigma2017, another option is to [run WireGuard Easy without Docker](/wg-easy/wg-easy/wiki/Using-WireGuard-Easy-without-Docker) directly on host.

**Reply by sigma2017** [reply]
Great to know but I don't know how.
Thank you.
LE: I will try, thanks so much.

**Reply by Semoz** [reply]
I also have the same requirement, which is to access the Caddy service after the client connects, hoping that Caddy can obtain the client's IP address and impose restrictions. 
After changing the wg-easy network to host mode (caddy is also in host mode), I did indeed obtain the client IP, which can access services such as ssh on the host machine, as well as services on other servers in the local area network.
 However, I cannot access the Docker service created by the bridge on this machine. 
wg-easy v15.3.0

**Comment by obuno** [comment]
Hi,

I'm posting here the modifications I've made in order NOT to have any WG_POST_UP nor WG_POST_DOWN on my `wg0` interface.
I have to specify also that I'm running wg-easy within an LXC container, hence no Docker. My early setup is close to this one although with modified IP schemes etc.. [https://github.com/wg-easy/wg-easy/wiki/Using-WireGuard-Easy-without-Docker](https://github.com/wg-easy/wg-easy/wiki/Using-WireGuard-Easy-without-Docker)

My `/etc/sysctl.conf` file host the following IPv4 parameters:
```
net.ipv4.ip_forward=1
net.ipv4.conf.all.src_valid_mark=1
```

Within my setup I wanted the WireGuard subnet/clients to be `routed` through the WireGuard host main interface, in my case, `eth0`. On the WireGuard host, a default gateway routing table entry is bound to `eth0` toward its connected subnet gateway. Adjacent gateway which knows where to route traffic toward the WireGuard subnet/clients. To reach the WireGuard host from public networks, such traffic would cross NGFW to reach the UDP:51820 destination onto the wg-easy enabled host on `eth0`.

In order to avoid both, WG_POST_UP and WG_POST_DOWN onto my `wg0.conf` interface settings, this is the file I've modified/edited: `/app/config.js` (save an original copy before making any changes...)

And here are the modified lines:
````
--- ORIGINAL /app/config.js.ORG FILE CONTENTS for WG_POST_UP BELOW ---

module.exports.WG_POST_UP = process.env.WG_POST_UP || `
iptables -t nat -A POSTROUTING -s ${module.exports.WG_DEFAULT_ADDRESS.replace('x', '0')}/24 -o ${module.exports.WG_DEVICE} -j MASQUERADE;
iptables -A INPUT -p udp -m udp --dport 51820 -j ACCEPT;
iptables -A FORWARD -i wg0 -j ACCEPT;
iptables -A FORWARD -o wg0 -j ACCEPT;
`.split('\n').join(' ');

--- EDITED /app/config.js FILE CONTENTS for WG_POST_UP BELOW ---

module.exports.WG_POST_UP = process.env.WG_POST_UP || '';


--- ORIGINAL /app/config.js.ORG FILE CONTENTS for WG_POST_DOWN BELOW ---

module.exports.WG_POST_DOWN = process.env.WG_POST_DOWN || `
iptables -t nat -D POSTROUTING -s ${module.exports.WG_DEFAULT_ADDRESS.replace('x', '0')}/24 -o ${module.exports.WG_DEVICE} -j MASQUERADE;
iptables -D INPUT -p udp -m udp --dport 51820 -j ACCEPT;
iptables -D FORWARD -i wg0 -j ACCEPT;
iptables -D FORWARD -o wg0 -j ACCEPT;
`.split('\n').join(' ');

--- EDITED /app/config.js FILE CONTENTS for WG_POST_DOWN BELOW ---

module.exports.WG_POST_DOWN = process.env.WG_POST_DOWN || '';

````

Obviously, I've got to save my edited `config.js` prior to any updates. In fact, in my books; NAT is useful at the boundary where you do not have control anymore. Which is not the case for me within my own networks. Also, this allows me with proper authentications onto further internal systems, proper activity trails and logs all bound to the WireGuard client IP and not the WireGuard host itself...

I'm not pretending that my config is 100% correct, I might have missed some needed details although it does function as I want it to.

Hope this helps,
Kind regards
